### PR TITLE
implemented non blocking login (fixes #38)

### LINKED ISSUE #38: Telegram.login() dont use stdin/stdout
Hi,
I am trying to build a telegram-client with this library, that does not get its inputs from stdin. Unfortunately `Telegram.login()` asks the confirmation code from the standard input. Is it possible to relay the code in any other matter? 
Like one call to start the process of obtaining the code, and a call to another function to verify it. This way it would allow me to forward the request to the UI. 
Another thing I noticed is that by using the library I have a lot of output in my stdout. This seems to be some kind of debug mode. Is it possible to run it outside of a debug mode?

### PR DESCRIPTION
**Pull request by melfnt**
I have implemented telegram client non-blocking mode login. For a concrete example, see the file examples/get_me_noblock_login.py

I've added an optional parameter ask_code_and_passwd to the login() function: if it is True (the default) the login process starts in blocking mode as it used to do.
If ask_code_and_passwd is False, whenever telegram code or password are required to complete the login process, the login() function returns with a meaningful return value.
The caller should ask the user for the telegram code and/or password depending on value returned by login() and then call the methods set_telegram_code() and/or set_telegram_password() and finally call login() again.

### CODE REVIEW

**Comment by melfnt** [comment]
I have added an example file but no unit test because I don't know how to test this feature.

**Comment by alexander-akhmetov** [comment]
Looks great, thank you! I've been thinking about this too and probably the login states can be also hidden if instead of `ask_code_and_passwd` we could pass two functions (one to get a password and one to get a code) to the `login` method:

```python
def get_code():
    return input('Please insert pin code here: ')

def get_password():
    return input('Insert password here: ')

tg = Telegram(...)
tg.login(
    get_code_func=get_code, 
    get_password_func=get_password,
)
```
And we can keep default blocking behaviour if `get_code_func` or `get_password_func` is not implemented. So, the login method could be called once and the Telegram instance calls relevant functions to get a code or a password. Do you mind to do the change? Otherwise, I can create another issue for this.

**Comment by melfnt** [comment]
Hi, I can do the change of course.

However, there are scenarios in which the user would prefer a true non-blocking approach rather than a different implementation of the get_code() and get_password() i.e. in some cases get_code() cannot be just a simple function call.

Consider for example the very asynchronous scenario in which the calls to the Telegram object are routed by requests to a REST server:
`
@app.route('/login methods=['POST'])
def login ( self ):
    self.tg = Telegram (...)
    self.tg.login ()

@app.route('/complete_login', methods=['POST'])
def complete_login ( self ):
    self.tg.set_code ( request.code )
    self.tg.set_password ( request.password )
    self.tg.login ()
`

In this case it is difficult (although possible) to implement the login process in a blocking fashion.

I agree that maybe we should hide the login state and maybe we can use two different functions: one to start the login process and the second one to complete it instead of calling login() twice.
Let me know what do you think.

**Comment by alexander-akhmetov** [comment]
> I agree that maybe we should hide the login state and maybe we can use two different functions: one to start the login process and the second one to complete it instead of calling login() twice.

That would be better, I think. We also can use the new functions in the `login` method. 

My main concern about the PR is that the `login` method should be called twice and we will have two similar methods to work with passwords and codes. We already have a set of private methods for different authorization states (`_send_telegram_code`, `_send_password`, etc.). They could be useful if we made them public to allow users just manage authorization states by themselves. 

The main idea of `login` method was to hide all the states to simplify the process, but I also wanted to make current private methods publicly available to use in more complicated use cases.

Would it solve your problem if the library had these public methods or/and functions to start, send code, send password and complete the login process?

**Comment by melfnt** [comment]
> That would be better, I think. We also can use the new functions in the `login` method.
>

I'll implement them soon, even if I think we should let the user choose between blocking mode (possibly with custom `get_code_func` and `get_passwd_func` functions) and pure non-blocking mode.
 
> My main concern about the PR is that the `login` method should be called twice and we will have two similar methods to work with passwords and codes. We already have a set of private methods for different authorization states (`_send_telegram_code`, `_send_password`, etc.). They could be useful if we made them public to allow users just manage authorization states by themselves.
> 
> The main idea of `login` method was to hide all the states to simplify the process, but I also wanted to make current private methods publicly available to use in more complicated use cases.
> 
> Would it solve your problem if the library had these public methods or/and functions to start, send code, send password and complete the login process?

Whether it solves or not my problem is a rather complicated question. I think I can explain myself better by having a closer look at the example `examples/get_me_noblock_login.py` I provided with the PR.

```
state = tg.login (ask_code_and_passwd=False)
if state == "authorizationStateWaitCode":
    # checkpoint (1)
    pin = input ("Please insert pin code here: ")
    tg.set_telegram_code (pin)
    state = tg.login(ask_code_and_passwd=False)
if state == "authorizationStateWaitPassword":
    # checkpoint (2)
    pwd = getpass.getpass('Insert password here (but please be sure that no one is spying on you): ')
    tg.set_telegram_password (pwd)
    tg.login ()
```

If the execution flow reaches the `checkpoint (1)`, I know for sure that the telegram client is blocked awaiting for a code. Thus, the next method call should be `_send_telegram_code()` (if it will be made public) rather than login() again.
Similarly, f the execution flow reaches the `checkpoint (2)` I know for sure that the next method call could be `_send_password()` rather than `login()` again.
So in a sense making `_send_telegram_code()` and `_send_password()` will solve my problem.

That said, the `login()` method  is a bit more powerful than the two methods `_send_telegram_code()` and `_send_password()` as it is albe to automatically resume the login process from the point it was previously stopped.
Let's make a simple example: if I already logged in from a device, telegram (and tdlib) do require to insert the password but do not require the telegram code again. That means means that if I run `examples/get_me_noblock_login.py` from a device where I have already logged in, the first `login()` invocation halts returning `'authorizationStateWaitPassword'`. Thus the `checkpoint (1)` is skipped and the password is required.
Thus the login state cannot be completely hidden to the caller (what should I ask next? The code or the password? What should I call next?) and calling `login()` or another member function is not very different.

I also don't think that asynchronous login is a complicated use case, it happens a lot when you are dealing with any kind of GUI.

To sum up I think it is better to call `login()` twice; or `login()` first and then other functions that sets the code/password and continue the login process rather than handling the whole login process manually calling `_send_telegram_code` and `_send_password()`.

I want also to hear @christofsteel opinion, as they are the related issue #38 opener.

**Comment by melfnt** [comment]
Just to be clear, I will be happy if the non-blocking login protocol of the telegram client is modified such that I can rewrite my example `examples/get_me_noblock_login.py` like this:
```
state = tg.login (ask_code_and_passwd=False)
if state == "authorizationStateWaitCode":
    # checkpoint (1)
    pin = input ("Please insert pin code here: ")
    state = tg.verify_code(pin)
if state == "authorizationStateWaitPassword":
    # checkpoint (2)
    pwd = getpass.getpass('Insert password here (but please be sure that no one is spying on you): ')
    tg.verify_password (pwd)
```

but ***ONLY IF*** the `verify_code()` and `verify_password()` functions are able to automatically resume the login process so that I won't have to handle it manually.
If this implies that `verify_code()` calls `login()` internally it is ok for me, but I think that it is not so much different than asking the caller to call `login()` twice.

**Comment by christofsteel** [comment]
So from a software design standpoint, I am not a fan of hiding a state if it exists. Each call to login should basically do the same thing. If something else should be done, another function should be called, or at least  the parameters should be different.
If I get this right, then `Telegram.login(ask_code_and_passwd=False)` does the following things:
 * if no pin information exists, the telegram servers send you a pin
 * if pin information exists, we login

Is it not possible to have a `Telegram.request_code` method, which does the first, and a `Telegram.login(pin=None)` method, which logs you in, with the given pin, and if no pin is given, it asks you via `input` for the pin. I am not familiar with passwords on Telegram, but I guess they can be handled in a similar fashion.

On another note: I would not use strings as markers for your state. It is easy for a developer to make a typo and never find the bug. Instead I would use enums, or if compatibility with python versions prior to 3.4 is an issue, aliases for integers.

**Comment by melfnt** [comment]
> So from a software design standpoint, I am not a fan of hiding a state if it exists. Each call to login should basically do the same thing. If something else should be done, another function should be called, or at least the parameters should be different.

I agree with you. The problem is that since the login state is managed by the telegram server each call to `login()` actually do always the same thing:
```
while not self._authenticated:
    fetch from server the login state
    do the next action
```
Please have a look to the current `login()` implementation.

> If I get this right, then `Telegram.login(ask_code_and_passwd=False)` does the following things:
> 
>     * if no pin information exists, the telegram servers send you a pin
> 
>     * if pin information exists, we login
> 
> 

That is correct.

> Is it not possible to have a `Telegram.request_code` method, which does the first, and a `Telegram.login(pin=None)` method, which logs you in, with the given pin, and if no pin is given, it asks you via `input` for the pin. I am not familiar with passwords on Telegram, but I guess they can be handled in a similar fashion.
> 

That is possible in principle, apart from the fact that in some cases the code is not required at all. This means the first function call must return a value that exposes the current login stats informing the caller whether code, password or neither of them is required.
Will the latest code snippet satisfy your requirements, replacing the name of the functions `login()`, `verify_code()` and `verify_password()` with more appropriate names?

> On another note: I would not use strings as markers for your state. It is easy for a developer to make a typo and never find the bug. Instead I would use enums, or if compatibility with python versions prior to 3.4 is an issue, aliases for integers.

I was using strings because these are the values used internally by tdlib. @alexander-akhmetov what do you think?

**Comment by alexander-akhmetov** [comment]
> I'll implement them soon, even if I think we should let the user choose between blocking mode (possibly with custom get_code_func and get_passwd_func functions) and pure non-blocking mode.

> but ONLY IF the verify_code() and verify_password() functions are able to automatically resume the login process so that I won't have to handle it manually.

That's almost what I mean by two functions.

We already have a state inside tdlib, we can provide low-level functions (which are private now) and they can be used to implement anything. 

To simplify things, two new functions can be used. We call a function to start the login process, then we send a code and/or a password and then we call another function to complete the process:

```python
# _set_initial_params, _send_encryption_key, _sent_phone_number_or_bot_token
state = tg.start_login()

if state == "authorizationStateWaitCode":
    pin = input ("pin:")
    state = tg.send_code(pin)  # blocking, returns authorization state

if state == "authorizationStateWaitPassword":
    pwd = getpass.getpass("password:")
    state = tg.send_password(pwd)    # blocking, returns authorization state

tg.complete_login()  # blocking, _complete_authorization
```

It's almost the same as your example, but doesn't save pin/password in the instance and uses the same mechanisms as we use right now to login. And the current `login()` method will behave as it is now. Though `send_password` and `send_code` should be both blocking. 

What do you think?

---

> On another note: I would not use strings as markers for your state. It is easy for a developer to make a typo and never find the bug. Instead I would use enums, or if compatibility with python versions prior to 3.4 is an issue, aliases for integers.

> I was using strings because these are the values used internally by tdlib. @alexander-akhmetov what do you think?

That would be great, but you can just keep them as strings if you want, and I'll create another issue to change this before releasing a new version.

**Comment by melfnt** [comment]
> > I'll implement them soon, even if I think we should let the user choose between blocking mode (possibly with custom get_code_func and get_passwd_func functions) and pure non-blocking mode.
> 
> > but ONLY IF the verify_code() and verify_password() functions are able to automatically resume the login process so that I won't have to handle it manually.
> 
> That's almost what I mean by two functions. Sorry, I didn't make myself clear.
> 
> We already have a state inside tdlib, we can provide low-level functions (which are private now) and they can be used to implement anything.
> 
> To simplify things, two new functions can be used (to start and to complete the process). We call a function to start the login process, then we send a code and/or a password and then we call another function to complete the process:
> 
> ```python
> # _set_initial_params, _send_encryption_key, _sent_phone_number_or_bot_token
> state = tg.start_login()
> 
> if state == "authorizationStateWaitCode":
>     pin = input ("pin:")
>     state = tg.send_code(pin)    # blocking, returns authorization state, calls _send_telegram_code(block=True) internally
> 
> if state == "authorizationStateWaitPassword":
>     pwd = getpass.getpass("password:")
>     state = tg.send_password(pwd)    # blocking, returns authorization state, calls _send_password(block=True)
> 
> tg.complete_login()  # blocking, _complete_authorization; or just login()
> ```
> 
> It's almost the same as your example, but doesn't save pin/password in the instance and uses the same mechanisms as we use right now to login. And the current `login()` method will behave as it is now. Though `send_password` and `send_code` should be both blocking.
> 
> What do you think?

That's perfect, I'd go for this implementation. 

> 
> > On another note: I would not use strings as markers for your state. It is easy for a developer to make a typo and never find the bug. Instead I would use enums, or if compatibility with python versions prior to 3.4 is an issue, aliases for integers.
> 
> > I was using strings because these are the values used internally by tdlib. @alexander-akhmetov what do you think?
> 
> That would be great, but you can just keep them as strings if you want, and I'll create another issue to change this before releasing a new version.

No I think we should not change the behaviour in a later version.  I will implement the solution that makes use of the enums from the very beginning

**Comment by melfnt** [comment]
Here I am again. This is the last interface I have implemented for the non-blocking login ( excerpt from file `examples/get_me_noblock_login` ):
```
state = tg.start_login()
if state == AuthorizationState.WAIT_CODE:
    pin = input("Please insert pin code here: ")
    state = tg.send_code(pin)
if state == AuthorizationState.WAIT_PWD:
    pwd = getpass.getpass('Insert password here (but please be sure that no one is spying on you): ')
    tg.send_password (pwd)
```

The functions `send_code()` and `send_password()` use `_send_telegram_code()` and `_send_password()`internally, plus a common helper method `_non_blocking_login_routine()`.
A final call to `tg.complete_login()` is not really needed because the process is completed by either of the three functions `start_login()`, `send_code(pin)` or `send_password (pwd)` when they attempt to get the next authorization state but the process is already complete.

I also didn't remove the old `login()` function, both for retrocompatibility and because I think it is really useful if one just wants to perform a blocking login operation.
So currently we have two entry functions:
```
login() #for blocking logins
start_login () #for non blocking logins, to be called in combination with send_code() and send_password()
```
That sounds a bit weird to me, I suggest to call them `login()` and `login_nowait()` (like the [Queue instance methods](https://docs.python.org/3.7/library/queue.html#queue.Queue.put_nowait)) or better, to introduce a parameter `block=True` and keep just `login` version.
We can also add the other two optional parameters `get_code_func` and `get_password_func` that @alexander-akhmetov  mentioned above.

I also need help to test the process a little bit more (as I said I couldn't add an automatic unit test) and with a pylint error:
```
telegram/client.py:25:0: C0103: Constant name "AuthorizationState" doesn't conform to UPPER_CASE naming style (invalid-name)
```
I am not aware that enums names should conform to UPPER_CASE naming style.
Furthermore, please fell free to change the name of the functions and/or the names of the enums memeber.

File: telegram/client.py [hunk 1]
```diff
@@ -21,6 +22,8 @@
 
 MESSAGE_HANDLER_TYPE: str = 'updateNewMessage'
 
+AuthorizationState = enum.Enum('AuthorizationState', 'IDLE WAIT_CODE WAIT_PWD DONE FAILED')
```
Comment (alexander-akhmetov) [inline telegram/client.py]: > I am not aware that enums names should conform to UPPER_CASE naming style.
```
telegram/client.py:25:0: C0103: Constant name "AuthorizationState" doesn't conform to UPPER_CASE naming style (invalid-name)
```

It is a convention for constants and in your case, the linter just sees it as a constant. You need to change it to something like this, and it can be reused in the login method:

```python
class AuthorizationState(enum.Enum):
    NOT_STARTED: str = None
    WAIT_TDLIB_PARAMETERS: str = 'authorizationStateWaitTdlibParameters'
    WAIT_ENCRYPTION_KEY: str = 'authorizationStateWaitEncryptionKey'
    WAIT_PHONE_NUMBER: str = 'authorizationStateWaitPhoneNumber'
    WAIT_CODE: str = 'authorizationStateWaitCode'
    WAIT_PASSWORD: str = 'authorizationStateWaitPassword'
    READY: str = 'authorizationStateReady'
```

File: telegram/client.py [hunk 2]
```diff
@@ -382,6 +386,20 @@ def get_authorization_state(self) -> AsyncResult:
 
         return self._send_data(data, result_id='getAuthorizationState')
 
+    def _wait_authorization_result(self, result) -> str:
+        authorization_state = None
+        if result:
```
Comment (alexander-akhmetov) [inline telegram/client.py]: When it can be empty/None?

File: telegram/client.py [hunk 3]
```diff
@@ -382,6 +386,20 @@ def get_authorization_state(self) -> AsyncResult:
...
+            result.wait(raise_exc=True)
```
Comment (alexander-akhmetov) [inline telegram/client.py]: I've added a new parameter to the `_send_data` function: `block`, it can be used in the `_send_password` and `_send_telegram_code` methods to prevent waiting here.

File: telegram/client.py [hunk 4]
```diff
@@ -382,6 +386,20 @@ def get_authorization_state(self) -> AsyncResult:
...
+
+            if result.update is None:
```
Comment (alexander-akhmetov) [inline telegram/client.py]: `result.update` should always be presented, there is no need for this check

File: telegram/client.py [hunk 5]
```diff
@@ -407,17 +425,55 @@ def login(self) -> None:
         while not self._authorized:
             logger.info('[login] current authorization state: %s', authorization_state)
             result = actions[authorization_state]()
+            authorization_state = self._wait_authorization_result(result)
+
+    def _non_blocking_login_routine(self, authorization_state) -> AuthorizationState:
```
Comment (alexander-akhmetov) [inline telegram/client.py]: Since we have authorization states as a part of the library's API, let's merge all these into login method with `block=True` as you suggested. If the `send_code` and `send_password` continue the login process, we definetely don't need the `start_login`. I'm not sure that they should do this, but I do agree that it's more convenient from the user's perspective.

So the `login(block=True)` method will perform all authorization steps and ask codes and passwords using stdin, and `login(block=False)` will return authorization state if it requires any blocking action (code or password in our case).

**Comment by alexander-akhmetov** [comment]
> introduce a parameter block=True and keep just login version.

Sounds good, let's do this. 

> We can also add the other two optional parameters get_code_func and get_password_func that @alexander-akhmetov mentioned above.

No, we don't need this if we have non-blocking login with `send_password` and `send_code` methods.

> I also need help to test the process a little bit more (as I said I couldn't add an automatic unit test) 

Don't worry, I will add new tests.

File: telegram/client.py (hunk 1 above)
Comment (melfnt) [reply telegram/client.py]: Roger that

File: telegram/client.py (hunk 2 above)
Comment (melfnt) [reply telegram/client.py]: I keep the original code from [python-telegram/client.py at line 418 on the master branch](https://github.com/alexander-akhmetov/python-telegram/blob/master/telegram/client.py#L418); you were checking for result not to be none and so did I.

Anyway, in that point of the code result is the return value of one of the actions in the actions dict:
```
{
            None: self.get_authorization_state,
            'authorizationStateWaitTdlibParameters': self._set_initial_params,
            'authorizationStateWaitEncryptionKey': self._send_encryption_key,
            'authorizationStateWaitPhoneNumber': self._send_phone_number_or_bot_token,
            'authorizationStateWaitCode': self._send_telegram_code,
            'authorizationStateWaitPassword': self._send_password,
            'authorizationStateReady': self._complete_authorization,
}
```
The last function `self._complete_authorization` [has no return statement](https://github.com/alexander-akhmetov/python-telegram/blob/master/telegram/client.py#L503), so that's where result can be none.

File: telegram/client.py (hunk 3 above)
Comment (melfnt) [reply telegram/client.py]: roger that

File: telegram/client.py (hunk 4 above)
Comment (melfnt) [reply telegram/client.py]: See the previous comment and [python-telegram/client.py at line 421 on branch master](https://github.com/alexander-akhmetov/python-telegram/blob/master/telegram/client.py#L421)

File: telegram/client.py (hunk 5 above)
Comment (melfnt) [reply telegram/client.py]: Perfect, I'll merge the two implementations

**Comment by melfnt** [comment]
Hi, sorry for the late replies, I was on holidays last week. I'll work on this in the next few days

**Comment by alexander-akhmetov** [comment]
Hi @melfnt, I can continue the work to get this merged if you don't mind

**Comment by melfnt** [comment]
> Hi @melfnt, I can continue the work to get this merged if you don't mind

Hi @alexander-akhmetov sorry for not being around such a long time but I had some trouble with other projects on September.
My app works just fine with the partial version of non blocking login I have already implemented, so my organization does not give me enough time to continue the work on python-telegram client. That said, I can resume my work next week if you have some other things to do.

I think it's a two days work or less, so I hope to finish before October the 18th. If it's too late for you please feel free to finish it by yourself.

Sorry again,
melfnt

**Comment by alexander-akhmetov** [comment]
@melfnt  It doesn't look like there is an urgency here, so it's totally fine, I just didn't want to continue working on your PR without asking you. Feel free to ping me if you don't have time to finish it or if you want me to continue working on this. Thanks for your contribution and all the suggestions!

**Comment by melfnt** [comment]
@alexander-akhmetov I tried to fix the issues you pointed out, but I was not able to get my code work with your new version of the `_send_data` method with `block=True`.
I don't want to push not working code and I won't be able to work on this for the next two weeks, so if you want you can continue.

I'm really sorry I couldn't make it.

**Comment by alexander-akhmetov** [comment]
@melfnt Thank you very much for contribution!

Do you mind if I continue based on your changes?

**Comment by melfnt** [comment]
Sure, no problem

**Review by ESTHERTERSEN — COMMENTED** [review]
Start2
